lib/adapters/http.js

```diff
@@ -13,33 +14,54 @@ var zlib = require('zlib');
 var VERSION = require('./../env/data').version;
 var createError = require('../core/createError');
 var enhanceError = require('../core/enhanceError');
-var defaults = require('../defaults');
 var Cancel = require('../cancel/Cancel');
+var defaults = require('../defaults');
 
 var isHttps = /https:?/;
 
 /**
  *
  * @param {http.ClientRequestArgs} options
- * @param {AxiosProxyConfig} proxy
+ * @param {AxiosProxyConfig} configProxy
  * @param {string} location
  */
-function setProxy(options, proxy, location) {
-  options.hostname = proxy.host;
-  options.host = proxy.host;
-  options.port = proxy.port;
-  options.path = location;
-
-  // Basic proxy authorization
-  if (proxy.auth) {
-    var base64 = Buffer.from(proxy.auth.username + ':' + proxy.auth.password, 'utf8').toString('base64');
-    options.headers['Proxy-Authorization'] = 'Basic ' + base64;
+function setProxy(options, configProxy, location) {
+  var proxy = configProxy;
+  if (!proxy && proxy !== false) {
+    var proxyUrl = getProxyForUrl(location);
+    if (proxyUrl) {
+      proxy = url.parse(proxyUrl);
+      // replace 'host' since the proxy object is not a URL object
+      proxy.host = proxy.hostname;
+    }
   }
+  if (proxy) {
+    // Basic proxy authorization
+    if (proxy.auth) {
+      // Support proxy auth object form
+      if (proxy.auth.username || proxy.auth.password) {
+        proxy.auth = (proxy.auth.username || '') + ':' + (proxy.auth.password || '');
+      }
+      var base64 = Buffer
+        .from(proxy.auth, 'utf8')
+        .toString('base64');
+      options.headers['Proxy-Authorization'] = 'Basic ' + base64;
+    }
 
-  // If a proxy is used, any redirects must also pass through the proxy
-  options.beforeRedirect = function beforeRedirect(redirection) {
-    redirection.headers.host = redirection.host;
-    setProxy(redirection, proxy, redirection.href);
+    options.headers.host = options.hostname + (options.port ? ':' + options.port : '');
```

Comment:
This `headers.host` assignment was done two different ways:
- When initializing the request: https://github.com/axios/axios/pull/4436/files?w=1#diff-586c04c24608b86a074073c5514cc195be5c4c87ddb56da233f6d998ada4159fL216
- When following redirects: https://github.com/axios/axios/pull/4436/files?w=1#diff-586c04c24608b86a074073c5514cc195be5c4c87ddb56da233f6d998ada4159fL41

I moved this here because it's now possible to have different effective proxying on redirects when re-resolving the proxy from environment variables - ie set for `http_proxy` but not set for `https_proxy`, or skipping due to `no_proxy` for one call but not a redirect, etc.

Reply:
Cool 😄